src/meta/src/barrier/mod.rs

```diff
@@ -254,106 +369,255 @@ where
+            tracing::info!(
+                "complete epoch{:?}",
+                command_ctx_queue_clone
+                    .get(0)
+                    .unwrap()
+                    .command_ctx
+                    .prev_epoch
+            );
+            let result = GlobalBarrierManager::try_commit_epoch(
+                &mut command_ctx_queue_clone,
+                barrier_send_tx_clone.clone(),
+                hummock_manager.clone(),
+                wait_build_actor.clone(),
+            )
+            .await;
+            if result.is_err() {
+                wait_recovery.store(true, atomic::Ordering::SeqCst);
+                command_ctx_queue_clone.get_mut(0).unwrap().states = Fail(result.unwrap_err());
+            }
+        }
+        if command_ctx_queue_clone.len() > 0
+            && matches!(command_ctx_queue_clone.get(0).unwrap().states, Fail(_))
+        {
+            let count = command_ctx_queue_clone
+                .iter()
+                .filter(|x| matches!(x.states, InFlight))
+                .count();
+            if count == 0 {
+                let new_epoch = command_ctx_queue_clone
+                    .get(command_ctx_queue_clone.len() - 1)
+                    .unwrap()
+                    .command_ctx
+                    .curr_epoch;
+
+                let mut err_msg = None;
+                while let Some(node) = command_ctx_queue_clone.pop_front() {
+                    let err = match node.states {
+                        Fail(err) => err,
+                        Complete => RwError::from(ErrorCode::InternalError(
+                            "err from before epoch".to_string(),
+                        )),
+                        _ => {
+                            panic!(
+                                "fail recovery with pending : epoch{:?}",
+                                node.command_ctx.prev_epoch
+                            );
                         }
+                    };
 
-                        state.prev_epoch = new_epoch;
-                    } else {
-                        panic!("failed to execute barrier: {:?}", e);
+                    node.timer.unwrap().observe_duration();
+                    node.notifiers
+                        .into_iter()
+                        .for_each(|notifier| notifier.notify_collection_failed(err.clone()));
+                    if err_msg.is_none() {
+                        err_msg = Some(err);
+                    }
+                    match node.command_ctx.command {
+                        Command::Plain(..) => {}
+                        _ => {
+                            wait_build_actor.store(false, atomic::Ordering::SeqCst);
+                        }
                     }
                 }
+                let err_msg = err_msg.unwrap();
+                barrier_send_tx_clone
+                    .send(BarrierSendResult::Err { new_epoch, err_msg })
+                    .unwrap();
```

Comment:
Did you try triggering the recovery process via manual trigger or via test? I suspect whether it works after checking the implementation of `GlobalBarrierManager::recovery`.